scripts/fast-reboot

```diff
@@ -72,42 +87,74 @@ docker kill lldp > /dev/null
 # Kill teamd, otherwise it gets down all LAGs
 docker kill teamd > /dev/null
 
+# Freeze orchagent for warm restart
+if [[ "$REBOOT_TYPE" = "warm-reboot" ]]; then
+    docker exec -i swss /usr/bin/orchagent_restart_check || {
+        echo "RESTARTCHECK failed\n" >&2
+        # TODO: remove this sleep if we have better timing control
+        sleep 1
+        exit 10
+    }
+fi
+
 # syncd graceful stop is supported only for Broadcom platforms only for now
-if [[ "$sonic_asic_type" = 'broadcom' ]];
-then
+if [[ "$sonic_asic_type" = 'broadcom' ]]; then
     # Gracefully stop syncd
-    docker exec -i syncd /usr/bin/syncd_request_shutdown --cold > /dev/null
+    if [[ "$REBOOT_TYPE" = "fast-reboot" ]]; then
+        docker exec -i syncd /usr/bin/syncd_request_shutdown --cold > /dev/null
+    elif [[ "$REBOOT_TYPE" = "warm-reboot" ]]; then
+        docker exec -i syncd /usr/bin/syncd_request_shutdown --warm > /dev/null
+    fi
 
     # Check that syncd was stopped
     while docker top syncd | grep -q /usr/bin/syncd
     do
-          sleep 0.1
+        sleep 0.1
     done
 fi
 
+# Kill swss and syncd dockers
+docker kill swss syncd
+
+# Warm reboot: dump state to host disk
+if [[ "$REBOOT_TYPE" = "warm-reboot" ]]; then
+    # Dump redis content to directory
+    # Note: don't use pretty mode redis-dump (1.1) since there is a known bug with key pattern
+    DUMP_CMD="redis-dump -s /var/run/redis/redis.sock"
+    WARM_DIR=/host/warm-reboot
+    mkdir -p $WARM_DIR
+    # Note: requiring redis-dump-load
+    # Save applDB in /host/warm-reboot/appl_db.json
+    $DUMP_CMD -d 0 -o $WARM_DIR/appl_db.json
+    # Save configDB in /host/warm-reboot/config_db.json
+    $DUMP_CMD -d 4 -o $WARM_DIR/config_db.json
+    # Save stateDB (only FDB_TABLE and WARM_RESTART_TABLE) in /host/warm-reboot/state_db.json
+    # WARNING WARNING WARNING: a trick to dump both FDB_TABLE|* and WARM_RESTA*
+    # TODO: replace it with readable mechanism to dump multiple key patterns into one single json file
+    $DUMP_CMD -d 6 -k "[FW][DA][BR][_M][T_][AR][BE][LS][ET][|A]*" -o $WARM_DIR/state_db.json
+    # Save asicDB in /host/warm-reboot/asic_db.json
+    $DUMP_CMD -d 1 -o $WARM_DIR/asic_db.json
```

Comment:
The create/remove/set from orchagent is asynchronous,  after reboot the transient requests will be applied to ASIC directly (but with the delay of whole reboot time window)?  After applying the delayed request, syncd performs further request from new orchagent?

I see the "delayed" direct ASIC operation could cause problem potentially.  it might mess up the warm restart processing.

Reply:
@yxieca I think syncd warm restart needs to take care of the possible transient data issue(before shutdown /after restart) , at least  to make sure it doesn't bring any side effect.

Reply:
The concern is valid and we should prevent it somewhere.